File: cmd/project_context.go
package cmd

import (
	"os"
	"strings"

	"github.com/pcr-developers/cli/internal/projects"
)

// projectContext holds the resolved project filter for the current directory.
type projectContext struct {
	// display name for the header (innermost matched project)
	name string
	// for filtering store queries
	ids   []string
	names []string
	// singleRepo is true when cwd is exactly a registered project (not a parent workspace).
	// In this mode, bundling is scoped strictly to this repo — cross-repo drafts remain
	// available for bundling from their other repos.
	singleRepo bool
}

// resolveProjectContext finds the registered project(s) relevant to the current directory.
//
// Exact match (cwd == registered project path): returns only that single project.
// This prevents parent workspaces from being pulled in when you're inside a specific repo.
//
// No exact match (e.g. running from pcr-developers/ parent folder): returns all
// registered projects nested under cwd, enabling workspace-wide bundling.
func resolveProjectContext() projectContext {
	cwd, _ := os.Getwd()
	if gitRoot := gitOutput("git", "rev-parse", "--show-toplevel"); gitRoot != "" {
		cwd = gitRoot
	}

	projs := projects.Load()

	// Exact match: running from within a specific registered project.
	// Scope strictly to that project — don't pull in parent workspaces.
	for _, p := range projs {
		if p.Path == cwd {
			ctx := projectContext{name: p.Name, singleRepo: true}
			if p.ProjectID != "" {
				ctx.ids = []string{p.ProjectID}
			}
			if p.Name != "" {
				ctx.names = []string{p.Name}
			}
			if p.ClaudeSlug != "" {
				ctx.names = append(ctx.names, p.ClaudeSlug)
			}
			return ctx
		}
	}

	// No exact match — include all registered projects nested under cwd
	// (handles parent workspace directories like pcr-developers/).
	seen := map[string]bool{}
	var ctx projectContext
	for _, p := range projs {
		if !strings.HasPrefix(p.Path, cwd+"/") {
			continue
		}
		if seen[p.Path] {
			continue
		}
		seen[p.Path] = true
		if ctx.name == "" || len(p.Path) > len(getProjectByName(projs, ctx.name).Path) {
			ctx.name = p.Name
		}
		if p.ProjectID != "" {
			ctx.ids = append(ctx.ids, p.ProjectID)
		}
		if p.Name != "" {
			ctx.names = append(ctx.names, p.Name)
		}
		if p.ClaudeSlug != "" {
			ctx.names = append(ctx.names, p.ClaudeSlug)
		}
	}
	return ctx
}

func getProjectByName(projs []projects.Project, name string) projects.Project {
	for _, p := range projs {
		if p.Name == name {
			return p
		}
	}
	return projects.Project{}
}

File: crates/pcr-core/src/commands/project_context.rs
//! Resolves the registered projects relevant to the current working
//! directory. Mirrors `cli/cmd/project_context.go` one-for-one.

use std::path::PathBuf;

use crate::projects::{self, Project};
use crate::sources::shared::git;

#[derive(Debug, Default, Clone)]
pub struct ProjectContext {
    /// Display name for the header (innermost matched project).
    pub name: String,
    /// Project IDs used when filtering store queries.
    pub ids: Vec<String>,
    /// Project names (and claude slugs) used when filtering store queries.
    pub names: Vec<String>,
    /// True when cwd is exactly a registered project path.
    pub single_repo: bool,
}

pub fn resolve() -> ProjectContext {
    let mut cwd = std::env::current_dir()
        .unwrap_or_else(|_| PathBuf::from("."))
        .to_string_lossy()
        .into_owned();
    let git_root = git::git_output(&["rev-parse", "--show-toplevel"]);
    if !git_root.is_empty() {
        cwd = git_root;
    }

    let projs = projects::load();

    // Exact match first — strict single-repo scoping.
    for p in &projs {
        if p.path == cwd {
            let mut ctx = ProjectContext {
                name: p.name.clone(),
                single_repo: true,
                ..Default::default()
            };
            if !p.project_id.is_empty() {
                ctx.ids = vec![p.project_id.clone()];
            }
            if !p.name.is_empty() {
                ctx.names.push(p.name.clone());
            }
            if !p.claude_slug.is_empty() {
                ctx.names.push(p.claude_slug.clone());
            }
            return ctx;
        }
    }

    let mut ctx = ProjectContext::default();
    let mut best_len = 0usize;
    let mut seen = std::collections::HashSet::<String>::new();
    for p in &projs {
        let prefix = format!("{}/", cwd);
        if !p.path.starts_with(&prefix) {
            continue;
        }
        if !seen.insert(p.path.clone()) {
            continue;
        }
        if p.path.len() > best_len {
            ctx.name = p.name.clone();
            best_len = p.path.len();
        }
        if !p.project_id.is_empty() {
            ctx.ids.push(p.project_id.clone());
        }
        if !p.name.is_empty() {
            ctx.names.push(p.name.clone());
        }
        if !p.claude_slug.is_empty() {
            ctx.names.push(p.claude_slug.clone());
        }
    }
    ctx
}

/// `projByID` lookup from [`cli/cmd/bundle.go::loadProjByID`].
pub fn load_proj_by_id() -> std::collections::BTreeMap<String, String> {
    let mut m = std::collections::BTreeMap::new();
    for p in projects::load() {
        if !p.project_id.is_empty() {
            m.insert(p.project_id, p.name);
        }
    }
    m
}

pub fn _unused_project(p: &Project) -> &Project {
    p
}
